### ISSUE TITLE
React wrapper can't set property correctly for lazy rendered rangepicker

### ISSUE THREAD

**Issue by shinxi** [PredixDev/px-rangepicker#58]
## Background

We've had some discussion at https://github.com/PredixDev/px-calendar-picker/issues/14#issuecomment-353161912 to use React refs to manually set some object type property of predix components. It works well for those components rendered at the first page load.

## Issues
But it can't work well for components that are lazy rendered, e.g, 2rd tab of a tabs component, we only can see time is settled, but date is not.

## Code example
https://codesandbox.io/s/1ow02wvv64

## Possible fix

Dive deeper, we've found for px-datetime-picker/px-datetime-field, the lazy-render issue possibly has fixed at px-datetime-field@2.0.4, it should be ok if upgrade to 2.0.4, however...bower doesn't have the latest 2.0.4 version. See https://github.com/PredixDev/px-datetime-field/issues/6.
But even after be able to upgrade to px-datetime-field, we'll still be facing lazy-render issues at px-rangepicker/px-datetime-range-field.

@katemenkhaus @davidrleonard @Aileen-H @mdwragg pls help take a look. Thanks in advance.
cc @shao @joshsylvester @inothings @elvinfucom

**Comment by mdwragg** [comment]
Wow, that's a lot of code.... I honestly don't have much of a clue what is going on here (I'm not that familiar with the date time stuff in the design system and React scares me...).

Genuine question (as I'm interested in the 'why'). What advantage does this 'wrapping in React' give you vs. using Polymer as the view construct and redux as the state/session store? Why did we choose the architecture? How is it helping us build faster etc?

The polymer + redux pattern has been talked about a lot, e.g. https://www.youtube.com/watch?v=PahsgJn0sgU

**Comment by davidrleonard** [comment]
So, after stepping through this code a bit I think we have an underlying timing issue in the px-rangepicker component that is causing this. Sometimes when the application sets the `fromMoment` it can get wiped out unexpectedly from inside `px-rangepicker`. 

This is a potential issue for any consumer and is not specific to React or "lazy loading" scenarios. I think the "lazy load" simulation here just exposes a different timing of component initialization then property setting that triggers this bug. 

We'll take a look at that underlying issue, once we have a fix we'll update this thread: #59

**Comment by shinxi** [comment]
> Genuine question (as I'm interested in the 'why'). What advantage does this 'wrapping in React' give you vs. using Polymer as the view construct and redux as the state/session store? Why did we choose the architecture? How is it helping us build faster etc?
> The polymer + redux pattern has been talked about a lot, e.g. https://www.youtube.com/watch?v=PahsgJn0sgU

@mdwragg First of all, we've already used redux at our React application, the code sample I'm showing is just for demo purpose, it's not production code. 
And for why not standalone Polymer application, that's a long story. Actually we've tried the standalone polymer before, however the performance is not good when loading large data on grid/tree, possibly due to polymer v1's performance issue. But the performance is just one point, we choose React not only because it works well with state management library redux, but more importantly React has a more mature ecosystem for development lifecycle. See [this presentation](https://shinxi.github.io/react-steps/#/1) for more detail if you are interested. :)
Giving the benefits from React, our architects finally decide to use React as our UI library. Also given by the situation that there are already many awesome predix-ui components, we decided to wrap these predix-ui components in React so that we can use them at our React application directly to save efforts.
